File: backend/src/main/kotlin/inzynierka/myhotelassistant/models/service/ServiceTypeAttributes.kt
package inzynierka.myhotelassistant.models.service

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes(
    JsonSubTypes.Type(ServiceTypeAttributes.Selection::class, name = "SELECTION"),
    JsonSubTypes.Type(ServiceTypeAttributes.Custom::class, name = "CUSTOM"),
)
sealed interface ServiceTypeAttributes {
    fun getType(): ServiceType

    data class OptionObject(
        val label: String,
        val description: String,
        val price: Double,
        val image: String? = null,
    )

    data class Selection(
        val multipleSelection: Boolean,
        val options: LinkedHashMap<String, List<OptionObject>>,
    ): ServiceTypeAttributes {
        override fun getType(): ServiceType = ServiceType.SELECTION
    }


    data class Custom(
        var fields: LinkedHashMap<String, Any>,
    ): ServiceTypeAttributes {
        override fun getType(): ServiceType = ServiceType.CUSTOM
    }
}

File: frontend/src/types/service_type_attributes.ts
type TypeProp = { type: string; }

export interface OptionObject {
  label: string;
  description: string;
  price: number;
  image?: string;
}

export interface SelectionAttributes extends TypeProp {
  multipleSelection: boolean;
  options: Record<string, OptionObject[]>;
}

export interface SelectionAttributesResponse extends TypeProp {
  selectedOptions: Record<string, number>;
}

export type ServiceDetails = SelectionAttributes;

export type ServiceDetailsResponse = SelectionAttributesResponse;
